contracts/cw3-fixed-multisig/src/contract.rs

```diff
@@ -0,0 +1,267 @@
+    latest: Option<Expiration>,
+) -> StdResult<HandleResponse<Empty>> {
+    // only members of the multisig can create a proposal
+    let raw_sender = deps.api.canonical_address(&env.message.sender)?;
+    let weight = voter_weight_read(&deps.storage)
+        .may_load(raw_sender.as_slice())?
+        .ok_or_else(StdError::unauthorized)?;
+
+    // TODO: using max as default here, also enforce max
+    let cfg = config_read(&deps.storage).load()?;
+    let expires = latest.unwrap_or(cfg.max_voting_period);
+
+    // create a proposal
+    let prop = Proposal {
+        title,
+        description,
+        expires,
+        msgs,
+        status: Status::Open,
+        yes_weight: weight,
+        required_weight: cfg.required_weight,
+    };
+
+    // get next id
+    let id = next_id(&mut deps.storage)?;
+
+    // save the proposal
+    proposal(&mut deps.storage).save(&id.to_be_bytes(), &prop)?;
+
+    // add the first yes vote from voter
+    let ballot = Ballot {
+        weight,
+        vote: Vote::Yes,
+    };
+    ballots(&mut deps.storage, id).save(raw_sender.as_slice(), &ballot)?;
+
+    // TODO: add some event attributes
+    Ok(HandleResponse::default())
+}
+
+pub fn handle_vote<S: Storage, A: Api, Q: Querier>(
+    deps: &mut Extern<S, A, Q>,
+    env: Env,
+    proposal_id: u64,
+    vote: Vote,
+) -> StdResult<HandleResponse<Empty>> {
+    // only members of the multisig can vote
+    let raw_sender = deps.api.canonical_address(&env.message.sender)?;
+    let weight = voter_weight_read(&deps.storage)
+        .may_load(raw_sender.as_slice())?
+        .ok_or_else(StdError::unauthorized)?;
+
+    // ensure proposal exists and can be voted on
+    let mut prop = proposal_read(&deps.storage).load(&proposal_id.to_be_bytes())?;
+    if prop.status != Status::Open {
+        return Err(StdError::generic_err("Proposal is not open for voting"));
+    }
+    if prop.expires.is_expired(&env.block) {
+        return Err(StdError::generic_err("Proposal voting period has expired"));
+    }
+
+    // cast vote if no vote previously cast
+    ballots(&mut deps.storage, proposal_id).update(raw_sender.as_slice(), |bal| match bal {
+        Some(_) => Err(StdError::generic_err("Already voted on this proposal")),
+        None => Ok(Ballot { weight, vote }),
+    })?;
+
+    // if yes vote, update tally
+    if vote == Vote::Yes {
+        prop.yes_weight += weight;
+        proposal(&mut deps.storage).save(&proposal_id.to_be_bytes(), &prop)?;
+    }
+
+    // TODO: add event attributes
+    Ok(HandleResponse::default())
+}
+
+pub fn handle_execute<S: Storage, A: Api, Q: Querier>(
+    deps: &mut Extern<S, A, Q>,
+    env: Env,
+    proposal_id: u64,
+) -> StdResult<HandleResponse<Empty>> {
+    // anyone can trigger this if the vote passed
+
+    let mut prop = proposal_read(&deps.storage).load(&proposal_id.to_be_bytes())?;
+    if prop.status != Status::Open {
+        return Err(StdError::generic_err("Proposal is not open for voting"));
```

Comment:
"Proposal is not open for voting" -> "Proposal is not open for execution" /  "Proposal is not open"?